### PR TITLE
CORE-4439 Gateway certificate validation

### PR DESCRIPTION
**Pull request by williamvigorr3**
Move the session certificate revocation check from the link manager to the gateway. Communication between the LinkManager and Gateway is done via the message bus. Using a RPCSender: RevocationCheckerClient in the LinkManager, as well as a RPCResponderProcessor: RevocationCheckProcessor in the Gateway. If in the future other components also want to delegate the revocation check to the Gateway, then we should pull out the RevocationCheckerClient into a common module. During session negotiation the Link Manager will make a request to the gateway.

Merge (https://github.com/corda/corda-api/pull/683) first.

### CODE REVIEW

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        private const val groupAndClientName = "GatewayRevocationChecker"
+        private const val certificateAlgorithm = "PKIX"
+        private const val certificateFactoryType = "X.509"
+
+        fun getCertCheckingParameters(trustStore: KeyStore, revocationConfig: RevocationConfig): PKIXBuilderParameters {
+            val pkixParams = PKIXBuilderParameters(trustStore, X509CertSelector())
+            val revocationChecker = when (revocationConfig.mode) {
+                RevocationConfigMode.OFF -> AllowAllRevocationChecker
+                RevocationConfigMode.SOFT_FAIL, RevocationConfigMode.HARD_FAIL -> {
+                    val certPathBuilder = CertPathBuilder.getInstance("PKIX")
```
Comment (dimosr) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt]:
```suggestion
                    val certPathBuilder = CertPathBuilder.getInstance(certificateAlgorithm)
```

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                ))
+                return
+            }
+            if (certificateChain == null) {
+                respFuture.completeExceptionally(
+                    IllegalStateException("Revocation check request cannot be made, certificate chain could not be parsed from pem format.")
+                )
+                return
+            }
+            val pkixRevocationChecker = getCertCheckingParameters(trustStore, revocationMode.toRevocationConfig())
```
Comment (dimosr) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt]:
```suggestion
            val pkixParams = getCertCheckingParameters(trustStore, revocationMode.toRevocationConfig())
```
nit

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+            try {
+                certPathValidator.validate(certificateChain, pkixRevocationChecker)
+            } catch (exception: CertPathValidatorException) {
+                respFuture.complete(RevocationCheckStatus.REVOKED)
```
Comment (dimosr) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt]: I think when revocation fails we get a quite human-friendly error message on the reason, right?

Should we add an extra field in the response to include that?

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                return
+            }
+            respFuture.complete(RevocationCheckStatus.ACTIVE)
+        }
+
+        private fun RevocationMode.toRevocationConfig(): RevocationConfig {
```
Comment (dimosr) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt:124]: should we change the `getCertCheckingParameters` to take `RevocationConfigMode` as parameter?

If we did that, I think we wouldn't need this.

File: libs/crypto/crypto-utils/src/main/kotlin/net/corda/crypto/utils/CertificateUtils.kt [hunk 5]
```diff
@@ -146,4 +146,9 @@ object AllowAllRevocationChecker : PKIXRevocationChecker() {
...
         return LinkedList()
     }
-}
\ No newline at end of file
+}
+
+data class KeyStoreWithPem(
+    val keyStore: KeyStore,
+    val pemKeyStore: List<PemCertificate>
+)
```
Comment (dimosr) [inline libs/crypto/crypto-utils/src/main/kotlin/net/corda/crypto/utils/CertificateUtils.kt]: These represent the same certificates, right? Is there a reason for passing them around together vs passing only one of the two formats and converting between them, where needed?

File: libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt [hunk 6]
```diff
@@ -74,23 +76,21 @@ class CertificateValidator(
...
-                }
-                pkixRevocationChecker
-            }
+    private fun validateRevocation(certificateChain: CertPath, pemCertificates: List<String>, trustStore: List<String>) {
+        val revocationMode = when (revocationCheckMode) {
+            RevocationCheckMode.OFF -> return //No check to do
+            RevocationCheckMode.HARD_FAIL -> RevocationMode.HARD_FAIL
+            RevocationCheckMode.SOFT_FAIL -> RevocationMode.SOFT_FAIL
+        }
+        val revocationStatus = checkRevocation(RevocationCheckRequest(pemCertificates, trustStore, revocationMode))
```
Comment (dimosr) [inline libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:91]: should we skip checking if the mode is `OFF`?

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/RevocationCheckerClient.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    private val rpcSender = RPCSenderWithDominoLogic(
+        publisherFactory, coordinatorFactory, publisherConfig, messagingConfiguration
+    )
+
+    fun checkRevocation(request: RevocationCheckRequest): RevocationCheckStatus {
+        return try {
+            rpcSender.sendRequest(request).getOrThrow(timeout)
+        } catch (except: TimeoutException) {
+            logger.warn("Revocation request sent to the gateway timed out.")
+            return RevocationCheckStatus.REVOKED
```
Comment (dimosr) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/RevocationCheckerClient.kt]: I believe `SOFT_FAIL` mode allows one to consider a certificate non-expired when revocation status check was impossible due to network errors.

Should we adjust the response accordingly here in that case?

And also, see my other comment about not sending requests when mode is `OFF`. If we send them, I think it's possible for them to timeout and then we return back `REVOKED` here, which wouldn't be ideal.

**Review by dimosr — COMMENTED** [review]
I did a first pass, leaving a few comments. And I'll come back to have a quick look at tests.

File: libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt (hunk 6 above)
Comment (williamvigorr3) [reply libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:91]: We are doing that (return on line 81).

File: libs/crypto/crypto-utils/src/main/kotlin/net/corda/crypto/utils/CertificateUtils.kt (hunk 5 above)
Comment (williamvigorr3) [reply libs/crypto/crypto-utils/src/main/kotlin/net/corda/crypto/utils/CertificateUtils.kt]: Two reasons: Firstly we will fail earlier if the trust store is invalid (when the group gets updated) (perhaps this is unlikely to happen because we hopefully validate this when the group policy is uploaded). Secondly if we store both formats we don't have to convert between the two (in the LinkManager) every time we do a revocation check.

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/RevocationCheckerClient.kt (hunk 7 above)
Comment (williamvigorr3) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/RevocationCheckerClient.kt]: I think if this request times out it means the link manager can't talk to the gateway through the message bus. If this is the case it is unlikely to be a situation we can recover from quickly even if the certificate is treated as ACTIVE.

File: libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt (hunk 6 above)
Comment (dimosr) [reply libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:91]: d'uh, my brain must have short-circuited the `when` clause waiting for an if-clause right after it.

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/RevocationCheckerClient.kt (hunk 7 above)
Comment (dimosr) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/RevocationCheckerClient.kt]: I think it doesn't necessarily mean that the link manager can't talk to the gateway at all. It is possible this request is not completed even though the link manager can transfer data to the gateway through other means fine (e.g. via the `link.out` topic), e.g. because there was some rebalance and the request got lost. In general, rpc request/reply pattern is expected to be unreliable.

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt (hunk 4 above)
Comment (williamvigorr3) [reply components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt:124]: We also use `getCertCheckingParameters` in the http client. So it takes the more extensive `RevocationConfigMode` which also includes the mode `OFF`. My opinion (weakly) favours leaving the code as is and so we don't have to duplicate some of the logic inside `getCertCheckingParameters`.

File: libs/crypto/crypto-utils/src/main/kotlin/net/corda/crypto/utils/CertificateUtils.kt (hunk 5 above)
Comment (dimosr) [reply libs/crypto/crypto-utils/src/main/kotlin/net/corda/crypto/utils/CertificateUtils.kt]: yes, I think we are safe on the first point. On the second point, won't the session manager end up performing a lookup every time, which internally performs a conversion anyway? Or do we use both, in which case we'd need extra conversions?

I am just wondering whether we are prematurely optimising here, because this might end up being a considerable amount of memory which might end up having adverse effects due to GC churn. So, in the absence of concrete data doing the simpler thing might be preferable.

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt (hunk 4 above)
Comment (dimosr) [reply components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt:124]: ah yes, I forgot that we have two different modes now. So, when commenting, I was thinking we could just do `revocationCheck.mode` at the invocation site in `HttpClient`. But, we'd have to do the conversions between the two types anyway.

File: components/gateway/src/integrationTest/kotlin/net/corda/p2p/gateway/certificates/RevocationCheckerTest.kt [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import org.junit.jupiter.api.assertThrows
+import java.time.Duration
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.TimeUnit
+
+class RevocationCheckerTest {
+    private companion object {
+        val futureTimeOut: Duration = Duration.ofSeconds(120)
+        val lifecycleCoordinatorFactory = LifecycleCoordinatorFactoryImpl(LifecycleRegistryImpl(), LifecycleCoordinatorSchedulerFactoryImpl())
+        val topicService = TopicServiceImpl()
```
Comment (dimosr) [inline components/gateway/src/integrationTest/kotlin/net/corda/p2p/gateway/certificates/RevocationCheckerTest.kt]: I think we should avoid adding more dependencies on the (deprecated) in-memory version, especially given we'll be removing it very soon.

File: components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/certificates/RevocationCheckerTest.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+                any(),
+                processor.capture(),
+            )
+        } doReturn subscription
+    }
+    private val mockDominoTile = Mockito.mockConstruction(RPCSubscriptionDominoTile::class.java) { _, context ->
+        @Suppress("UNCHECKED_CAST")
+        (context.arguments()[1] as  () -> RPCSubscription<RevocationCheckRequest, RevocationCheckStatus>)()
+    }
+    private val revocationChecker = RevocationChecker(subscriptionFactory, mock(), mock())
```
Comment (dimosr) [inline components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/certificates/RevocationCheckerTest.kt]: I think this is unused? And if you remove it, there are a couple of other things that are unused and can be removed e.g. the subscription.

File: libs/p2p-crypto/src/test/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidatorTest.kt [hunk 10]
```diff
@@ -46,15 +51,19 @@ class CertificateValidatorTest {
 
     @Test
     fun `certificate fails validation if X500 name doesn't match`() {
-        val validator = CertificateValidator(RevocationCheckMode.HARD_FAIL, trustStore, certPathValidator, certificateFactory)
+        val validator = CertificateValidator(
```
Comment (dimosr) [inline libs/p2p-crypto/src/test/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidatorTest.kt:59]: should we now add a unit test for the scenario where revocation checks are enabled and the check returns negative (certificate is revoked)?

And _maybe_ one when the mode is OFF, where regardless of whether it's revoked validation succeeds (because no check is performed)?

Comment (williamvigorr3) [reply libs/p2p-crypto/src/test/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidatorTest.kt:59]: We integration test these scenarios at the moment. We probably can unit test them instead, to do that we would probably need to mock the construction of PKIXBuilderParameters. Do you think it is worth it?

Comment (dimosr) [reply libs/p2p-crypto/src/test/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidatorTest.kt:59]: ~We could also maintain the integration test nature of it, but switch to using the actual in-memory/DB version of the library.~ scratch that, I thought your response was to another comment. If it's covered by integration tests, it should be ok.

File: components/gateway/src/integrationTest/kotlin/net/corda/p2p/gateway/certificates/RevocationCheckerTest.kt (hunk 8 above)
Comment (dimosr) [reply components/gateway/src/integrationTest/kotlin/net/corda/p2p/gateway/certificates/RevocationCheckerTest.kt]: One more think about this before I go on PTO.. I was thinking that it might be better to integration test the message bus as part of the message bus library once, instead of every component integration testing it.

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt (hunk 3 above)
Comment (williamvigorr3) [reply components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt]: I did this and reworked the response.

File: libs/p2p-crypto/src/test/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidatorTest.kt (hunk 10 above)
Comment (williamvigorr3) [reply libs/p2p-crypto/src/test/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidatorTest.kt:59]: This is tested in the other `CertificateValidatorTest`.

File: components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/certificates/RevocationCheckerTest.kt (hunk 9 above)
Comment (williamvigorr3) [reply components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/certificates/RevocationCheckerTest.kt]: Only the `val` is unused. I moved it into an init block.

File: libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt [hunk 11]
```diff
@@ -83,8 +84,17 @@ class CertificateValidator(
...
+        if (status is Revoked) {
+            val x509CertChain = certificateChain.toX509()
+            if (status.certificateIndex >= 0 && status.certificateIndex < x509CertChain.size) {
+                certificateChain.toX509()[status.certificateIndex]
+            } else {
+                null
+            }?. let { specificCert ->
+                throw InvalidPeerCertificate("The certificate failed a revocation check (in the gateway): " + status.reason, specificCert)
+            }
+            throw InvalidPeerCertificate("The certificate failed a revocation check (in the gateway): " + status.reason, x509CertChain)
```
Comment (dimosr) [inline libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt]: In theory, this library is supposed to be agnostic to higher-levels. So, it might be better to remove the `(in the gateway)` bit from the  message.

File: libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt [hunk 12]
```diff
@@ -34,11 +34,15 @@ class CertificateValidator(
...
-            }
-        })
+        val certificateChain = try {
+            certificateFactory.generateCertPath(pemCertificateChain.map { pemCertificate ->
+                ByteArrayInputStream(pemCertificate.toByteArray()).use {
+                    certificateFactory.generateCertificate(it)
+                }
+            })
+        } catch (except: CertificateException) {
+            throw InvalidPeerCertificate("Error passing the certificate from a pem file: \n" + except.message)
```
Comment (dimosr) [inline libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt]: Did you mean `parsing` (instead of `passing` here)?

File: libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt [hunk 13]
```diff
@@ -74,23 +81,32 @@ class CertificateValidator(
...
+            throw InvalidPeerCertificate("The certificate failed a revocation check (in the gateway): " + status.reason, x509CertChain)
         }
-        pkixParams.addCertPathChecker(revocationChecker)
+    }
 
+    private fun validateCertPath(certificateChain: CertPath) {
+        val trustStore = convertToKeyStore(certificateFactory, pemTrustStore, "session")
+            ?: throw InvalidPeerCertificate("Could not validate certificate, the trust store could not be read from a pem file.")
+        val pkixParams = PKIXBuilderParameters(trustStore, X509CertSelector())
+        pkixParams.addCertPathChecker(AllowAllRevocationChecker)
```
Comment (dimosr) [inline libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:110]: can we add a one-line comment here to assist people understand why we do this, e.g.
```
// revocation checks are performed separately (see checkRevocation() function), here we just validate the certificate chain. 
```

**Review by dimosr — COMMENTED** [review]
LGTM - just a few very minor comments

File: libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt (hunk 12 above)
Comment (williamvigorr3) [reply libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt]: 🤦

**Review by dimosr — DISMISSED** [review]

**Review by dimosr — APPROVED** [review]
